eclipse-jdt/eclipse.jdt.ui
# Improve the Safety of "Extract Local Variable" refactorings by identifying statements that may change the value of the extracted expressions #432

Here is the PR based on the previous discussion in the [issue-432](https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/432). 

## What it does
`ChangedValueChecker` is the core part that is used to identify statements that may change the value of the extracted expressions. It first records the values that the selected expression depends on during the initialization phase. Then it calculates the path between the temp declaration and the last candidate extracted expression. Lastly, it validates intermediate statements in parallel to detect the intersection of the value updated by the intermediate code and the value read by the expression without duplication.

For `ExtractTempRefactoring.java`, in addition to the call of the `ChangedValueChecker` class, we have changed the original refactoring strategy when replacing the selected expression with temp declaration. Considering such example:
```java
1 {
2   f();
3   int x=f();
4 }
5 {
6   int y=f();
7 }
```
If we select `f()` at Line 2, the refactored code shown as follows has obviously syntax errors: 
```java
1 {
2   int f=f();
3   int x=f;
4 }
5 {
6   int y=f;
7 }
```
I think it is better to extract only once for the selected expression used as a statement. Because such an expression tends to modify the system state (otherwise it would be useless).

## How to test
Tests have been appended in `ExtractTempTests` to verify the correctness of the patch. Test-140 is added to test for a variable the selected expression depends on has been changed within the method invocation. Test-141/Test-142 is to test the case of loop/condition statement to ensure that the path is not just calculated on the basis of offsets. Test-143, Test-144, and Test-145 were designed to test complex scenarios to ensure the robustness of the code concerning calculating the path. Test-146 and Test-147 were added to verify the identification of changed values under different cases. 

Notably, we disabled Test-46, Test-110, and Test-111 because of the changed refactoring strategy. Besides, Test-73 was also excluded due to the added more precise flow analysis.

## Author checklist
